### PR TITLE
Add docs for docker for mac + windows and change port config

### PR DESCRIPTION
**Pull request by muffinresearch**
* This adds docs for docker-for-mac (and docker-for-windows although I've not tested the windows docker solution with our instructions - we should look to do that or invite contributions to fix the docs as necessary) 
* ~~I've changed the port configuration to facilitate overrides. Unfortunately the `docker-compose.overrides.yml` files does not override settings that are a list (it adds to them) so you can't just simply change the ports and leave the defaults.  This is a breaking change and will require people with an existing setup to copy the ``docker-compose.overrides.yml`` into place from the template in the repo (the file itself is in .gitignore) to get the same default ports or change them - we'll need to communicate that on the mailing list when this lands.~~

I've now changed this to not require the copy of a template for the ports. Instead the common case is supported out of the box and if you need to customise the ports you'll have to run a longer command specifying the config override you need to use.

### CODE REVIEW

**Comment by wagnerand** [comment]
lgtm

File: docs/topics/install/docker.rst [hunk 1]
```diff
@@ -10,83 +10,78 @@ development environment.
...
+most minimal set of moving parts.
 
-    docker-machine create --driver=virtualbox addons-dev
+If you have problems, due to not meeting the requirenents or you're on windows
+and want to keep running docker-machine vms then docker-machine will still
+work just fine. See the docs for creating the vm here :ref:`creating-the-docker-vm`
 
-Then you can export the variables so that docker-compose can talk to
-the docker service. This command will tell you how to do that::
+.. note:: if you're on OSX and already have a working docker-machine setup you
```
Comment (tofumatt) [inline docs/topics/install/docker.rst]: While it's technically "OS X", I'd change this to "Mac" or "MacOS" as they are getting away from the OS X branding anyway and "Mac" stands out more.

File: docs/topics/install/docker.rst [hunk 2]
```diff
@@ -10,83 +10,78 @@ development environment.
...
-    docker-machine env addons-dev
+Setting up the containers
+~~~~~~~~~~~~~~~~~~~~~~~~~
 
-On a mac it's a case of running::
+.. note::
 
-    eval $(docker-machine env addons-dev)
+    docker-toolbox, docker-for-mac and docker-for-windows will install ``docker-compose``
+    for you. If you're on linux and you need it, you can install it manually with::
```
Comment (tofumatt) [inline docs/topics/install/docker.rst]: Should be "Linux" not "linux".

File: docs/topics/install/docker.rst [hunk 3]
```diff
@@ -10,83 +10,78 @@ development environment.
...
-   files.
-
-Accessing docker
-~~~~~~~~~~~~~~~~
-
-The last step is to grab the ip of the vm. If you're using docker-machine,
-you can get the ip like so::
-
-    docker-machine ip addons-dev
+   Generally docker requires the code checkout to exist within your home directory so
```
Comment (tofumatt) [inline docs/topics/install/docker.rst]: Docker is a proper noun so should be capitalised everywhere.

File: docs/topics/install/docker.rst [hunk 4]
```diff
@@ -10,83 +10,78 @@ development environment.
...
-    docker-toolbox_ for more info.
+Accessing the web server
+~~~~~~~~~~~~~~~~~~~~~~~~
 
-Now you can connect to port 80 of that ip address. Here's an example
-(your ip might be different)::
+By default our docker-compose config exposes the web-server on port 80 of localhost.
 
-    http://192.168.99.100/
+We use olympia.dev as the default hostname to access your container server (e.g. for
```
Comment (tofumatt) [inline docs/topics/install/docker.rst]: I'd wrap "olympia.dev" in `<code>` so it stands out as a URL here.

File: docs/topics/install/docker.rst [hunk 5]
```diff
@@ -10,83 +10,78 @@ development environment.
...
+Firefox Accounts). To be able access the development environment using http://olympia.dev
```
Comment (tofumatt) [inline docs/topics/install/docker.rst]: Could even make `olympia.dev` a link here so they can click it once finished 😄

File: docs/topics/install/docker.rst [hunk 6]
```diff
@@ -140,6 +135,81 @@ migrations::
...
 
+Here's a list of a few of the issues you might face when using docker.
+
+Can't access the web server?
+----------------------------
+
+Check you've created an hosts file entry pointing ``olympia.dev`` to the
+relevant ip address.
+
+Also make sure you've copied ``docker-compose.override.yml.tmpl`` to
```
Comment (tofumatt) [inline docs/topics/install/docker.rst]: Is there no way around this? It takes away from the turn-key `docker-compose pull && docker-compose up -d` setup. I didn't need to do this on my Mac and I hit `127.0.0.1` just fine...

File: docs/topics/install/docker.rst [hunk 7]
```diff
@@ -140,6 +135,81 @@ migrations::
...
+``docker-compose.override.yml`` to get the default ports. If you haven't
```
Comment (tofumatt) [inline docs/topics/install/docker.rst]: "If you haven't" should be followed by a comma or colon, for clarity here.

File: docs/topics/install/docker.rst [hunk 8]
```diff
@@ -140,6 +135,81 @@ migrations::
...
+stop the containers with ``docker-compose stop`` copy the file and restart
+with ``docker-compose up -d``.
+
+Another tip is to use ``docker-compose ps`` to check the status of the
```
Comment (tofumatt) [inline docs/topics/install/docker.rst]: The next sentence starts with "another" as well, I'd change this to simple say:

```
If containers are failing to start, use ``docker-compose ps`` to check the status of the containers.
```

File: docs/topics/install/docker.rst [hunk 9]
```diff
@@ -140,6 +135,81 @@ migrations::
...
+containers. If they are failing to start you should be able to tell here.
+
+Another way to find out what's wrong is to run ``docker-compose logs``.
+
+Getting "Programming error [table] doesn't exist"?
+--------------------------------------------------
+
+Check you've run the ``make initialize_docker`` step.
```
Comment (tofumatt) [inline docs/topics/install/docker.rst]: I would rephrase to `Make sure you've run the... []`. You can't really check (aside from remembering that you did), but if I'm troublingshooting and something say "make sure you run X" I'll just run it which is the goal of this 😄 .

File: docs/topics/install/docker.rst [hunk 10]
```diff
@@ -10,83 +10,78 @@ development environment.
 First you'll need to install docker_, please check the information for
 the installation steps specific to your operating system.
 
-.. note::
-    Docker recommends installing docker-toolbox_ if you're on OS X or
-    windows and that will provide you with the ``docker-machine`` and
-    ``docker-compose`` (Mac-only).
-
+There are generally two options for running docker depending on the platform
```
Comment (tofumatt) [inline docs/topics/install/docker.rst]: "generally" isn't needed here.

File: docs/topics/install/docker.rst [hunk 11]
```diff
@@ -10,83 +10,78 @@ development environment.
...
+you are running.
 
-.. _creating-the-docker-vm:
+ * Run docker on the host machine directly (recommended if supported)
```
Comment (tofumatt) [inline docs/topics/install/docker.rst]: It's basically always supported, with the advent of Docker for Windows/Mac, so "recommended" seems sufficient.

File: docs/topics/install/docker.rst [hunk 12]
```diff
@@ -10,83 +10,78 @@ development environment.
...
+ * Run docker-machine which will run docker inside a virtual-machine
 
-Creating the docker vm (mac/windows)
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Historically mac and windows could only run docker via a vm. That has
+recently changed with the arrival of docker-for-mac_ and docker-for-windows_.
 
-If you go the docker-machine route your first step is to create a vm::
+If your platform can run docker directly either on Linux, with docker-for-mac_
+or docker-for-windows_ then this is the easiest way to install docker with the
```
Comment (tofumatt) [inline docs/topics/install/docker.rst]: "with the most minimal set of moving parts" is a bit of a turn of phrase that will likely confuse ESL speakers and doesn't add much clarity for English-speakers, so this could be shortened.

File: docs/topics/install/docker.rst [hunk 13]
```diff
@@ -10,83 +10,78 @@ development environment.
...
+most minimal set of moving parts.
 
-    docker-machine create --driver=virtualbox addons-dev
+If you have problems, due to not meeting the requirenents or you're on windows
```
Comment (tofumatt) [inline docs/topics/install/docker.rst]: Typo ("requirenents" -> "requirements"). Windows should be capitalised here.

Also: which requirements?

File: docs/topics/install/docker.rst [hunk 14]
```diff
@@ -10,83 +10,78 @@ development environment.
...
+and want to keep running docker-machine vms then docker-machine will still
+work just fine. See the docs for creating the vm here :ref:`creating-the-docker-vm`
 
-Then you can export the variables so that docker-compose can talk to
-the docker service. This command will tell you how to do that::
+.. note:: if you're on OSX and already have a working docker-machine setup you
+   can run that and docker-for-mac (*but not docker-for-windows*) side by side.
+   The only caveat here  is that it's recommended that you keep the versions of
```
Comment (tofumatt) [inline docs/topics/install/docker.rst]: "here" could be removed (it doesn't add clarity) plus it has two spaces after it 😉

File: docs/topics/install/docker.rst [hunk 15]
```diff
@@ -10,83 +10,78 @@ development environment.
...
-    addresses can change over time. You can find out which IP is in use using
-    ``docker-machine ip [machine name]``
+    [ip-address]  olympia.dev
 
-You may need to use a reliable hostname to access your container server (e.g. for
-Firefox Accounts). You can set one by editing your ``/etc/hosts`` file on your
-native operating system. For example::
-
-    192.168.99.100  olympia.dev
+Typically the ip address is localhost 127.0.0.1 if you're using docker-machine
```
Comment (tofumatt) [inline docs/topics/install/docker.rst]: IP should be capitalised, I thought.

File: docs/topics/install/docker.rst [hunk 16]
```diff
@@ -140,6 +135,81 @@ migrations::
...
+Getting "Programming error [table] doesn't exist"?
+--------------------------------------------------
+
+Check you've run the ``make initialize_docker`` step.
+
+
+Port collisions (nginx container fails to start)
+------------------------------------------------
+
+Since by default the docker-compose file exposes the port to the nginx
```
Comment (tofumatt) [inline docs/topics/install/docker.rst]: This sentence is a bit hard to follow. What about:

```
If you're already running a service on port 80 on your host machine, the `nginx` container will fail to start. This is because the docker-compose file tells `nginx` to listen on port 80 by default.
```

File: docs/topics/install/docker.rst [hunk 17]
```diff
@@ -140,6 +135,81 @@ migrations::
...
+port 80.
+
+This problem will manifest itself by the services failing to start, you'll
+be able to see the error like so::
+
+    ERROR: for nginx  Cannot start service nginx:.....
+    ...Error starting userland proxy: Bind for 0.0.0.0:80: unexpected error (Failure EADDRINUSE)
+    ERROR: Encountered errors while bringing up the project.
+
+There's a couple of ways to fix it. Simple one is to find out what's running on
```
Comment (tofumatt) [inline docs/topics/install/docker.rst]: For some people (maybe who have a MAMP or something similar installed that already runs on port 80), this might not be the simple way to open up port 80, and might only work until a restart (when whatever was running there will start again). Plus giving a `sudo` command as the simple solution (especially when they'll likely need to `sudo kill` a process running on port 80) seems bad.

Giving them a way to change the port first seems best.

**Comment by tofumatt** [comment]
There are a fair amount of copy edits worth doing, as often once these docs are written they are left alone, so might as well get them now. 😄

**Comment by muffinresearch** [comment]
@tofumatt all good points, thanks for the suggested edits! I'll get those fixed up.

File: docs/topics/install/docker.rst (hunk 13 above)
Comment (muffinresearch) [inline docs/topics/install/docker.rst]: I was thinking of the requirements for windows for mac which are quite strict.

File: docs/topics/install/docker.rst (hunk 6 above)
Comment (muffinresearch) [inline docs/topics/install/docker.rst]: Ok, there is another way to do this. It will make it more incovenient if you want to change the ports but will mean the default requires no extra changes. I'll update the docs to reflect that way since I think it will be nice to not require a breaking change.

Comment (tofumatt) [inline docs/topics/install/docker.rst]: Yeah, I’d go for the easiest defaults first; if they have to change
something (which I’d venture most shouldn’t have to) it’s alright if it’s
more work as long as we explain it :-)

File: docs/topics/install/docker.rst [hunk 18]
```diff
@@ -1,111 +1,103 @@
...
+ * Run docker on the host machine directly (recommended)
+ * Run docker-machine which will run docker inside a virtual-machine
 
-If you go the docker-machine route your first step is to create a vm::
+Historically Mac and Windows could only run Docker via a vm. That has
+recently changed with the arrival of docker-for-mac_ and docker-for-windows_.
 
-    docker-machine create --driver=virtualbox addons-dev
+If your platform can run Docker directly either on Linux, with docker-for-mac_
+or docker-for-windows_ then this is the easiest way to run it the most minimal
```
Comment (tofumatt) [inline docs/topics/install/docker.rst]: "then this is the easiest way to run it the most minimal set of requirements"

I think this sentence got tweaked but the tweaking wasn't finished 😄 

"then this is the easiest way to run addons-server" would work.

File: docs/topics/install/docker.rst [hunk 19]
```diff
@@ -1,111 +1,103 @@
...
+set of requirements.
 
-Then you can export the variables so that docker-compose can talk to
-the docker service. This command will tell you how to do that::
+If you have problems, due to not meeting the minimum specifications for
+docker-for-windows_ or you'd prefer to keep running docker-machine vms then
```
Comment (tofumatt) [inline docs/topics/install/docker.rst]: This reads a lot better, thanks 👍

File: docs/topics/install/docker.rst [hunk 20]
```diff
@@ -1,111 +1,103 @@
...
-   your home directory into the docker vm. Your code checkout will need
-   to be within your home directory so that docker-compose will find the necessary
-   files.
-
-Accessing docker
-~~~~~~~~~~~~~~~~
-
-The last step is to grab the ip of the vm. If you're using docker-machine,
-you can get the ip like so::
+   Docker requires the code checkout to exist within your home directory so
```
Comment (tofumatt) [inline docs/topics/install/docker.rst]: Huh, I didn't know that, neat!

File: docs/topics/install/docker.rst [hunk 21]
```diff
@@ -1,111 +1,103 @@
...
+   that docker can mount the source-code into the container.
```
Comment (tofumatt) [inline docs/topics/install/docker.rst]: Docker here should be uppercase.

File: docs/topics/install/docker.rst [hunk 22]
```diff
@@ -140,6 +132,77 @@ migrations::
 Gotchas!
 ~~~~~~~~
 
+Here's a list of a few of the issues you might face when using Docker.
+
+Can't access the web server?
+----------------------------
+
+Check you've created an hosts file entry pointing ``olympia.dev`` to the
```
Comment (tofumatt) [inline docs/topics/install/docker.rst]: "an hosts file" should be "a host file"

File: docs/topics/install/docker.rst [hunk 23]
```diff
@@ -140,6 +132,77 @@ migrations::
...
+relevant ip address.
```
Comment (tofumatt) [inline docs/topics/install/docker.rst]: IP should be capitalised here too.

File: docs/topics/install/docker.rst [hunk 24]
```diff
@@ -140,6 +132,77 @@ migrations::
...
+
+If containers are failing to start  use ``docker-compose ps`` to check their
```
Comment (tofumatt) [inline docs/topics/install/docker.rst]: This hardly matters but there's two spaces after "start" 😉

**Comment by tofumatt** [comment]
r+wc

Awesome, thanks so much.
